google-cloud-firestore/test/google/cloud/firestore/bulkwriter_test.rb

```diff
@@ -0,0 +1,333 @@
+    response = batch_write_pass_resp 1
+    write_requests = [Google::Cloud::Firestore::Convert.write_for_create("#{documents_path}/cities/NYC", { foo: "bar"})]
+    request = batch_write_args write_requests
+
+    firestore_mock.expect :batch_write, response, request
+    result = bulk_writer.create "cities/NYC", { foo: "bar"}
+    bulk_writer.flush
+
+    firestore_mock.expect :batch_write, response, request
+    result_1 = bulk_writer.create "cities/NYC", { foo: "bar"}
+    bulk_writer.flush
+
+    _(result.value).must_be_kind_of Google::Cloud::Firestore::BulkWriterOperation::WriteResult
+    _(result_1.value).must_be_kind_of Google::Cloud::Firestore::BulkWriterOperation::WriteResult
+  end
+
+  it "cannot add operations after bulkwriter is closed" do
+    bw = firestore.bulk_writer
+    bw.close
+
+    error = assert_raises Google::Cloud::Firestore::BulkWriterError do
+      bw.create "cities/NYC", { foo: "bar"}
+    end
+    _(error.message).must_equal "BulkWriterError : Not accepting responses for now. Either closed or in flush state"
+  end
+
+  it "cannot add operations when bulk writer is in closed state" do
+    write_requests = [Google::Cloud::Firestore::Convert.write_for_create("#{documents_path}/cities/NYC", { foo: "bar"})]
+    request = batch_write_args write_requests
+    responses = [batch_write_fail_resp(1), batch_write_pass_resp(1)]
+    requests = [request] * responses.length
+    stub = BatchWriteStub.new responses, requests
+    firestore.service.instance_variable_set :@firestore, stub
+
+    bw = firestore.bulk_writer
+
+    bw.create "cities/NYC", { foo: "bar"}
+
+    Concurrent::Promises.future do
+      bw.flush
+    end
+
+    sleep 0.1
+    error = assert_raises Google::Cloud::Firestore::BulkWriterError do
+      bw.create "cities/NYC", { foo: "bar"}
+    end
+    _(error.message).must_equal "BulkWriterError : Not accepting responses for now. Either closed or in flush state"
+  end
+
+  describe "retry tests" do
+    it "do not exceed the max retry attempts" do
+      write_requests = [Google::Cloud::Firestore::Convert.write_for_create("#{documents_path}/cities/NYC", { foo: "bar"})]
+      request = batch_write_args write_requests
+      responses = [batch_write_fail_resp(1), batch_write_fail_resp(1), batch_write_fail_resp(1), batch_write_pass_resp(1)]
+      requests = [request] * responses.length
+      stub = BatchWriteStub.new responses, requests
+      firestore.service.instance_variable_set :@firestore, stub
+
+      bw = firestore.bulk_writer retries: 2
+
+      result = bw.create "cities/NYC", { foo: "bar"}
+      bw.flush
+      bw.close
+
+      assert_raises Google::Cloud::Firestore::BulkWriterException do
+        result.wait!
+      end
+      _(result.rejected?).must_equal true
+      _(result.reason).must_be_kind_of Google::Cloud::Firestore::BulkWriterException
+      _(result.reason.status).must_equal 1
+      _(result.reason.message).must_equal "Mock rejection"
```

Comment:
Maybe override the mock message in this case to actually show that the error is because of retry exceeding and not a random error. There is a chance of flakiness here.

Reply:
Personalized the failure message for each failure and updated accordingly.